### DISCUSSION TITLE
OAuth: Shutting down the warp server afterwards?

### DISCUSSION THREAD

**Discussion by panssystem** [sreeise/graph-rs-sdk#449] [Q&A]
Only vaguely related to this library, but maybe someone here will have solved it. I'm using code derived from the oauth_code_grant example, and it works, and I can move forward, but it bugs me that there's a web server sitting there using resources but otherwise doing nothing. I've tried using bind_with_graceful_shutdown, but I haven't figured out how to send a signal from the handle_redirect method. Anyone have any suggestions?

**Comment by sreeise** [comment]
Hey @panssystem

Have you tried a oneshot channel along with the bind_with_graceful_shutdown? 

Or possibly [Tokio signal ctrl_c](https://docs.rs/tokio/1.34.0/tokio/signal/index.html) and listening for the ctrl_c in bind_with_graceful_shudtown?

Look at the second response using tokio ctrl_c on this page: https://stackoverflow.com/questions/71258916/how-to-gracefully-shutdown-a-warp-server

I havn't tried it myself but it may work for what your wanting.

**Reply by sreeise** [reply]
I also don't know what type of setup your running, but im working on some updates right now to do interactive authentication using webview which will be shutdown automatically upon successful auth or upon the window closing of course. But still a few weeks to possibly a month from being merged.

**Reply by panssystem** [reply]
Ah, thanks. I'll probably wait for the update to finalize anything, but it's actually just a pet project to help me learn Rust and enhance some aspects of Microsoft Todo.

My most recent thought was to spin authentication into its own process, either via a separate program, or just command line arguments that just auth and quit, so that can work in the meantime.

**Reply by sreeise** [reply]
No problem. Its an annoying issue and since its just used in examples its more or less just not something Ive looked into that much. Your actually not the first person to ask this same question and maybe one day the examples will be updated to use something that can be shutdown easily. But yeah not a bad idea to just spin it off into another process you can shutdown.
